architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+- Support efficient querying of resource type data for any API consumer.
+- Register resource types for existing Radius functionality like Deployment Engine and `Applications.*` resource types.
+
+### Non goals
+
+- (out of scope): Complete the user-defined type implementation. This design document is scoped to registration of resource types.
+- (out of scope): Tooling for resource type registration (CLI, Bicep). This work is tracked by another design document.
+- (non-goal): Provider non-user-facing mechanisms for resource type registration.
+
+### User scenarios (optional)
+
+#### Defining new types
+
+Suzy is a platform engineer at Contoso working on the internal developer platform. She defines a resource type for an internal service `ContosoBus` named `Contoso.Platform/contosoBuses`. 
+
+She can use tools like Bicep, GitOps, or the `rad` CLI to register the new resource types in any tenant of Radius.
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+#### Removing default resource types
+
+Suzy is a platform engineer at Contoso working on the internal developer platform. She wants to avoid confusion about the set of resource types supported by the platform.
+
+The application teams are Contoso are allowed to use `Applications.Datastores/redisCaches` and `Applications.Datastores/postgreSQLDatabases`. Suzy uses the `rad` CLI to unregister other types in the `Applications.Datastores` namespace.
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+## User Experience (if applicable)
+
+See API design. Tooling and user-experiences are out of scope for this document.
+
+## Design
+
+### High Level Design
+
+The resource type registration APIs allow a user to control the set of resource types known to UCP. Since UCP is responsible for routing, it needs to know the set of resource providers and resource types so that it can send them traffic. 
+
+#### Resource Types
+
+The resource type registration APIs add two new resource types:
+
+- `System.Resources/resourceProviders` for resource providers.
+- `System.Resources/resourceProviders/resourceTypes` for resource types.
+
+These new resource types define the full CRUDL set of APIs. These APIs are optimized for operations that need to modify or deeply understand resource types. 
+
+`resourceTypes` is our first nested resource type in UCP/Radius. This is a good usecase for a nested resource types because:
+
+- The `resourceTypes` body will be be large due to the inclusion of an OpenAPI schema. 
+- Users need granular registration and unregistration of individual resource types. (See scenario 2 above).
+
+---
+
+These new resource types are scoped to the Radius plane. 
+
+> eg: `/planes/radius/local/providers/System.Resources/resourceProviders`
+
+The rationale for this is to support registration/unregistration per-tenant of Radius rather than per-installation.
```

Comment:
Would users want more granular control? Was this vetted with our partners yet?

Reply:
Let's discuss in the meeting and I'll post conclusions here.

Reply:
We've got strong validation on the idea of registering/unregistering at the plane scope. 

The idea of registering/unregistering at the resource group scope is pretty complex. It would be hard for us to get feedback on this right now. Just explaining the idea to most users would be complex.